src/services.h

```diff
@@ -636,67 +648,65 @@ namespace dd
      * @param sname service name
      * @param out output data object
      */
-    int predict(const APIData &ad, const std::string &sname, APIData &out,
-                const bool &chain = false)
+    int predict(const APIData &ad_in, const std::string &sname,
+                APIData &ad_out, const bool &chain = false)
     {
       std::chrono::time_point<std::chrono::system_clock> tstart
           = std::chrono::system_clock::now();
-      visitor_predict vp;
-      vp._ad = ad;
-      vp._chain = chain;
-      output pout;
+
+      int status = 0;
       auto llog = spdlog::get(sname);
       try
         {
           auto hit = get_service_it(sname);
-          pout = mapbox::util::apply_visitor(vp, (*hit).second);
+          auto &mllib = (*hit).second;
+          status = visitor_mllib::predict_job(mllib, ad_in, ad_out, chain);
         }
       catch (InputConnectorBadParamException &e)
         {
           llog->error("mllib bad param: {}", e.what());
-          pout._status = -2;
+          status = -2;
           throw;
         }
       catch (MLLibBadParamException &e)
         {
           llog->error("mllib bad param: {}", e.what());
-          pout._status = -2;
+          status = -2;
           throw;
         }
       catch (MLLibInternalException &e)
         {
           llog->error("mllib internal error: {}", e.what());
-          pout._status = -1;
+          status = -1;
           throw;
         }
       catch (MLServiceLockException &e)
         {
           llog->error("mllib lock error: {}", e.what());
-          pout._status = -3;
+          status = -3;
           throw;
         }
       catch (const std::exception &e)
         {
           // catch anything thrown within try block that derives from
           // std::exception
           llog->error("other error: {}", e.what());
-          pout._status = -1;
+          status = -1;
           throw;
         }
       catch (...)
         {
           llog->error("prediction call failed");
-          pout._status = -1;
+          status = -1;
           throw;
         }
-      out = pout._out;
```

Comment:
It seems like `out` could be modified by the predict call, this PR seems to remove this behaviour (`out` is passed by value to `predict_job()` and the modifications are discarded with the visitor).